utils/s3/client.cc

```diff
@@ -854,6 +864,238 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece));
 }
 
+// unlike upload_sink and upload_jumbo_sink, do_upload_file reads from the
+// specified file, and sends the data read from disk right away to the wire,
+// without accumulating them first.
+class client::do_upload_file {
+    shared_ptr<client> _client;
+    const std::filesystem::path _path;
+    sstring _object_name;
+    sstring _upload_id;
+    utils::chunked_vector<sstring> _part_etags;
+    std::optional<tag> _tag;
+    size_t _part_size;
+    gate _bg_uploads;
+
+    // each time, we read up to transmit size from disk.
+    // this is also an option which limits the number of multipart upload tasks
+    static constexpr size_t _transmit_size = 4_KiB;
```

Comment:
That's an argument to copy_to() call, so the copy_to() does
```
for (;;) {
    buf = in.read_up_to(4k);
    out.write(buf);
}
```
i.e. -- it reads 4k from disk, then puts 4k into network.
For output it's OK, because the buffer size defaults to default_body_sink_buffer_size which is ~32k
For file it results in `std::min(std::max(_options.buffer_size / 4, size_t(8192)), _options.buffer_size);` = 8k, I guess it's too low for throughput-oriented IO

Reply:
changed to 8KiB.

Reply:
That's not enough. The file input streams in the uploading sink are created with default options that are

```c++
struct file_input_stream_options {
    size_t buffer_size = 8192;    ///< I/O buffer size
    unsigned read_ahead = 0;      ///< Maximum number of extra read-ahead operations
    lw_shared_ptr<file_input_stream_history> dynamic_adjustments = { }; ///< Input stream history, if null dynamic adjustments are disabled
};      
```

i.e. -- still 8k reads from disk, no read-ahead, no on-demand input buffer growth

Reply:
this is exactly why i am using 8K now -- to match with the underlying buffer size. do you think we could have a chunk sizing greater than 8k when reading from disk?

Reply:
How much do we wish to send to S3 at a time, no? I know some clients use 64K or more. Minimizing the per-request overhead is also important, vs. parallel (and multi-part) uploads/downloads.

Reply:
> How much do we wish to send to S3 at a time, no? 

could you define "a time"? we are writing to the body of a tcp request. 

> I know some clients use 64K or more. Minimizing the per-request overhead is also important, 

and what does "request" stand for? we are sending one POST http request for each part, which sizes 5MiB by default.

> vs. parallel (and multi-part) uploads/downloads.

i am confused. so what do you suggest exactly?

Reply:
I assume we upload/download in parallel. 8K is too small, IMHO. Would be interesting to see what https://github.com/peak/s5cmd uses.

Reply:
s3cmd uses aws-sdk-go under the hood, see https://github.com/peak/s5cmd/blob/89218dc7fd484030367b3139961ba1be061d8b2b/storage/s3.go#L29-L32 and https://github.com/peak/s5cmd/blob/89218dc7fd484030367b3139961ba1be061d8b2b/storage/s3.go#L882

so i am reading aws-sdk-go for some inspirations. see https://github.com/aws/aws-sdk-go/blob/main/service/s3/s3manager/read_seeker_write_to.go#L41, which is using 64k. while aws-sdk-cpp uses the part size , see https://github.com/aws/aws-sdk-cpp/blob/11af3c647e84697a649351934060aa58663f3d59/src/aws-cpp-sdk-transfer/source/transfer/TransferManager.cpp#L452 .

so let's be more conservative and use 64k.

Reply:
> this is exactly why i am using 8K now -- to match with the underlying buffer size. do you think we could have a chunk sizing greater than 8k when reading from disk?

This default of 8k is aimed at "sstable query reader" case, i.e. -- when the read from an sstable is for a single entry. When issuing lots of small requests, disk will bottleneck at its IOPS, will provide nice single-request latency, but won't achieve the maximum throughput. Compaction (and AFAIK range reads) configure input stream the other way -- their use larger buffer size _and_ setup the dynamic adjustments, so that input stream starts reading data with larger chunks (up to 128k). With _that_ buffer size disk would bottleneck at its bandwidth, the single-request latency won't be as nice, IOPS will be low. But large bandwidth (not large IOPS) is what we need to upload a file

Reply:
64k underlying buffer size might not be enough either. iotune has the "saturation length" parameter, it's the buffer size using which one can achieve maximum throughput. On some disks it can be up to 1Mb.

Reply:
Another (minor) factor is to reduce the no. of requests, due to costs. See https://aws.amazon.com/s3/pricing/ . I think it's not a big deal, but would be good to understand it. You pay $0.005 for every 1000 PUT requests.

Reply:
> Another (minor) factor is to reduce the no. of requests, due to costs.

It doesn't directly depend on the IO size we're discussing here. Current implementation splits a file into uploading parts and each part is a single request. Smaller IO size makes scylla execute this request in several writes, but it's still single request.

On the other hand, the less requests we do, the longer each request lasts. In the extreme case we could upload a file in a single request. But the longer request lasts the higher probability of server connection drop is. And we cannot re-upload only the remainder of the part that failed, we need to re-upload the whole part.

Reply:
I don't think (but unsure) that multi-part upload is a single request (see https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html#mpuploadpricing - but it's not very clear there either).
Multi-part does allow retry of partial parts, I believe.

Reply:
Yes, it's not single request. The sequence is
* create (1 request)
* put part (N requests, file size = sum of par sizes, number of parts is N)
* complete (1 request)

My point is that each "put part" is a single request regardless of the IO size we use to read data from the file

My other point is that if N is large -> the part is small -> the chance of server connection closure is lower -> we likely don't re-upload the part. Contrary to this if N is small -> the part is large -> the chance of server connection closure is higher -> we likely will need to re-upload _the whole part_

Reply:
Multipart != multiple connections, AFAIK. It just allows to upload multiple parts in parallel over a single HTTP (TCP) connection, I believe. So the chances are not different for a connection closure? The part anyway should be small enough to indeed not suffer from this too much. I assume we'll know more by real life tests.

Reply:
Multipart is _exactly_ about multiple connections. One creates the upload context on the server side, gets ID for it, then PUTs as many parts as it wants identifying the destination with the upload-id. In parallel using as many TCP connections as client ~~wants~~ is able to maintain.

Multiple HTTP requests in parallel over a single TCP connection is impossible, the best HTTP protocol allows is pipelining, but it's not true parallelism, it's start writing new request without waiting for previous one's response